### PR TITLE
Simple Way to Parallel Verification Transaction

### PR DESCRIPTION
**Pull request by eryeer**
Here is a more simple way to parallel verification transaction to improve tx process performance.

Function `PipeTo` in Akka Framework can collect parallel execution result and send them to Akka Actor sequentially. We can use this way to parellelly verify Tx witness first and then send Tx to other sequential checks when blockchain receives a new Tx.

### CODE REVIEW

**Comment by erikzhang** [comment]
If we are under attack, a large number of transactions are sent from a certain node. Parallel verification of transactions in `Blockchain` will result in fewer opportunities for other normal user transactions to be verified. So I prefer to verify transactions in `ProtocolHandler` so that transactions sent by different nodes will not affect each other.

**Comment by Tommo-L** [comment]
> If we are under attack, a large number of transactions are sent from a certain node. ... so that transactions sent by different nodes will not affect each other.
 
We are with you for the security mechanism to limit attacks by malicious nodes, such as junk transactions, junk blocks, and massive transactions, etc. 

But we think it may be better to  put the security mechanism on the periphery of the execution process for the sake of efficiency, since it can lower the efficiency if we put  these security mechanisms in the execution process.

So I prefer to have another PR to address the security issue.

**Comment by vncoelho** [comment]
Such a good investigation @eryeer and @Tommo-L, congratulations for all the @neo-project/ngd-shanghai. 

It is an example of how we have potential for parallel processing on NEO, even GPU, which is a topic we research here in Brazil.

I also agree with @erikzhang, because the code is also current parallel. Thus, it would invert the processing power.
Let's keep this PR for not let the idea go away, however, let's not give full priority on this until we have a background scenario for detailed benchmarks.

**Comment by erikzhang** [comment]
I reorganized the code. It looks more clear.

**Comment by erikzhang** [comment]
Now it needs to test.

File: src/neo/Ledger/Blockchain.cs [hunk 1]
```diff
@@ -395,22 +397,43 @@ private void OnNewHeaders(Header[] headers)
...
+            else if (!transaction.Reverify(currentSnapshot, MemPool.SendersFeeMonitor.GetSenderFee(transaction.Sender)))
+                reason = RelayResultReason.Invalid;
+            else if (!NativeContract.Policy.CheckPolicy(transaction, currentSnapshot))
+                reason = RelayResultReason.PolicyFail;
+            if (reason != RelayResultReason.Succeed)
+            {
+                Sender.Tell(reason);
+                return;
+            }
+            Task.Run(() =>
```
Comment (vncoelho) [inline src/neo/Ledger/Blockchain.cs]: This @erikzhang always do magic! aehuaheauhuea

It looks quite nice organized like this, makes a lot of sense.

File: src/neo/Network/P2P/Payloads/Transaction.cs [hunk 2]
```diff
@@ -210,6 +210,11 @@ bool IInventory.Verify(StoreView snapshot)
         public virtual bool Verify(StoreView snapshot, BigInteger totalSenderFeeFromPool)
```
Comment (vncoelho) [inline src/neo/Network/P2P/Payloads/Transaction.cs]: How about modify its use on `private void OnFillMemoryPool(IEnumerable<Transaction> transactions)`?

Comment (vncoelho) [reply src/neo/Network/P2P/Payloads/Transaction.cs]: No need, I just checked `OnFillMemoryPool` is an isolated case only. Not worth.

**Comment by eryeer** [comment]
We did functional and performance tests, including normal transactions, duplicated transactions, and transactions with wrong signatures. The test results were in line with our expectations.

**Comment by shargon** [comment]
Much clearer 👍

**Comment by erikzhang** [comment]
Should we rename `Reverify()` to `VerifyForEachBlock()`?

**Comment by vncoelho** [comment]
`VerifyAgainstPool`?

**Comment by shargon** [comment]
@erikzhang Yes, `Reverify` seems worse. `VerifyForEachBlock` it's better.

**Comment by vncoelho** [comment]
It is not always `block`, sometimes it is against a mempool or a set created by the Primary.
But looks good as well yo be `VerifyForEachBlock`

**Comment by shargon** [comment]
Wait me for some test!

**Comment by shargon** [comment]
With 1m of invalid tx

```csharp
private Transaction CreateTransaction()
        {
            Random random = new Random();
            var randomBytes = new byte[16];
            random.NextBytes(randomBytes);

            return new Transaction()
            {
                Attributes = new TransactionAttribute[0],
                Cosigners = new Cosigner[0],
                NetworkFee = 100_000_000,
                Nonce = (uint)random.Next(),
                Script = randomBytes,
                ValidUntilBlock = 1000,
                Sender = UInt160.Zero,
                SystemFee = 10_000_000,
                Witnesses = new Witness[0]
            };
        }

        [TestMethod]
        public void TestTransaction()
        {
            var senderProbe = CreateTestProbe();
            var tx = CreateTransaction();

            var sw = System.Diagnostics.Stopwatch.StartNew();

            for (int x = 0; x < 1_000_000; x++)
            {
                senderProbe.Send(system.Blockchain, tx);
                senderProbe.ExpectMsg(RelayResultReason.Invalid);
            }

            sw.Stop();
            Console.WriteLine(sw.Elapsed);
        }
```

Before this patch:

```
00:00:05.7117458
```

After this patch
```
00:00:05.9725854
```

Seems that there are no difference. I am doing something wrong @eryeer ?

**Comment by vncoelho** [comment]
@shargon, try with a real ` Witnesses = new Witness[0]`.

**Comment by shargon** [comment]
@vncoelho If it's parallel, I think that it should be "almost" the same result with a invalid one

**Comment by vncoelho** [comment]
Exactly, "almost", that is why it is a little bit slower :dagger: 
Let's see the result with the real witness. Better to generate different witness for every tx.

**Comment by vncoelho** [comment]
@shargon, we also need the speed test also to return `RelayResultReason.Succeed` in order to ensure that all steps are executed on all threads.

**Comment by shargon** [comment]
Yes, you are right xD

**Comment by shargon** [comment]
The test was repeated :)

```csharp

        [TestMethod]
        public void TestValidTransaction()
        {
            var senderProbe = CreateTestProbe();

            var snapshot = Blockchain.Singleton.GetSnapshot();
            var walletA = TestUtils.GenerateTestWallet();
            using (var unlockA = walletA.Unlock("123"))
            {
                var acc = walletA.CreateAccount();

                // Fake balance

                var key = NativeContract.GAS.CreateStorageKey(20, acc.ScriptHash);
                var entry = snapshot.Storages.GetAndChange(key, () => new StorageItem
                {
                    Value = new Nep5AccountState().ToByteArray()
                });

                entry.Value = new Nep5AccountState()
                {
                    Balance = 100_000_000 * NativeContract.GAS.Factor
                }
                .ToByteArray();

                snapshot.Commit();

                typeof(Blockchain)
                    .GetMethod("UpdateCurrentSnapshot", BindingFlags.Instance | BindingFlags.NonPublic)
                    .Invoke(Blockchain.Singleton, null);

                // Make transaction
                var st = Stopwatch.StartNew();

                for (uint x = 0; x < 10_000; x++)
                {
                    var tx = CreateValidTx(walletA, acc.ScriptHash, x);
                    senderProbe.Send(system.Blockchain, tx);
                    senderProbe.ExpectMsg(RelayResultReason.Succeed);
                }

                st.Stop();
                Console.WriteLine(st.Elapsed);
            }
        }

        private Transaction CreateValidTx(NEP6Wallet wallet, UInt160 account, uint nonce)
        {
            var tx = wallet.MakeTransaction(new TransferOutput[]
                {
                    new TransferOutput()
                    {
                            AssetId = NativeContract.GAS.Hash,
                            ScriptHash = account,
                            Value = new BigDecimal(1,8)
                    }
                },
                account);

            tx.Nonce = nonce;

            var data = new ContractParametersContext(tx);
            Assert.IsTrue(wallet.Sign(data));
            Assert.IsTrue(data.Completed);

            tx.Witnesses = data.GetWitnesses();

            return tx;
        }
```
UT: eryeer#18

Results for 10K valid tx:

With this patch:

```
00:00:28.1237619
```

Without the patch:
```
00:00:35.4632680
```

So there was an improve :)
Good job @eryeer !

**Comment by shargon** [comment]
I think that we should move this checks to the top
```
BigInteger fee = SystemFee + NetworkFee + totalSenderFeeFromPool;
            if (balance < fee) return false;
            if (hashes.Length != Witnesses.Length) return false;
```
Because this was faster than the others.

**Comment by erikzhang** [comment]
https://github.com/neo-project/neo/blob/081dac5c763430cbb5672cc27581de8177c0f2ac/src/neo/Ledger/Blockchain.cs#L409

https://github.com/neo-project/neo/blob/081dac5c763430cbb5672cc27581de8177c0f2ac/src/neo/Network/P2P/Payloads/Transaction.cs#L277

I think they are duplicate checks too.

**Comment by shargon** [comment]
Yes, it is
```cshap
internal bool CheckPolicy(Transaction tx, StoreView snapshot)
        {
            UInt160[] blockedAccounts = GetBlockedAccounts(snapshot);
            if (blockedAccounts.Intersect(tx.GetScriptHashesForVerifying(snapshot)).Any())
                return false;
            return true;
        }
```

**Pull request merged**